src/connectors/impls/gpubsub/producer.rs

```diff
@@ -0,0 +1,298 @@
+        {
+            skip_authentication = self.config.skip_authentication;
+        }
+
+        let client = if skip_authentication {
+            info!("Skipping auth...");
+
+            PublisherClient::with_interceptor(
+                channel,
+                AuthInterceptor {
+                    token: Box::new(|| Ok(Arc::new(String::new()))),
+                },
+            )
+        } else {
+            let token = Token::new()?;
+
+            PublisherClient::with_interceptor(
+                channel,
+                AuthInterceptor {
+                    token: Box::new(move || {
+                        token.header_value().map_err(|_| {
+                            Status::unavailable("Failed to retrieve authentication token.")
+                        })
+                    }),
+                },
+            )
+        };
+
+        self.client = Some(client);
+
+        Ok(true)
+    }
+
+    async fn on_event(
+        &mut self,
+        _input: &str,
+        event: Event,
+        ctx: &SinkContext,
+        serializer: &mut EventSerializer,
+        _start: u64,
+    ) -> Result<SinkReply> {
+        let client = self.client.as_mut().ok_or(ErrorKind::ClientNotAvailable(
+            "PubSub",
+            "The publisher is not connected",
+        ))?;
+
+        let mut messages = Vec::with_capacity(event.len());
+
+        for (value, meta) in event.value_meta_iter() {
+            for payload in serializer.serialize(value, event.ingest_ns)? {
+                let ordering_key = ctx
+                    .extract_meta(meta)
+                    .get("ordering_key")
+                    .as_str()
+                    .map_or_else(|| "".to_string(), ToString::to_string);
+
+                messages.push(PubsubMessage {
+                    data: payload,
+                    attributes: HashMap::new(),
+                    // publish_time and message_id will be ignored in the request and set by server
+                    message_id: "".to_string(),
+                    publish_time: None,
+
+                    // from the metadata
+                    ordering_key,
+                });
+            }
+        }
+
+        if let Ok(inner_result) = client
+            .publish(PublishRequest {
+                topic: self.config.topic.clone(),
+                messages,
+            })
+            .timeout(Duration::from_nanos(self.config.request_timeout))
+            .await
+        {
+            if let Err(error) = inner_result {
+                error!("Failed to publish a message: {}", error);
+
+                if [
+                    Code::Aborted,
+                    Code::Cancelled,
+                    Code::DataLoss,
+                    Code::DeadlineExceeded,
+                    Code::Internal,
+                    Code::ResourceExhausted,
+                    Code::Unavailable,
+                    Code::Unknown,
+                ]
+                .contains(&error.code())
```

Comment:
We don't need to perform an array traversal for this

```suggestion
                if matches!(error.code(),
                    Code::Aborted |
                    Code::Cancelled |
                    Code::DataLoss |
                    Code::DeadlineExceeded |
                    Code::Internal |
                    Code::ResourceExhausted |
                    Code::Unavailable |
                    Code::Unknown
                )
```